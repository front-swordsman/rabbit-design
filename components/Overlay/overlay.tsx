import * as React from 'react'
import * as ReactDOM from 'react-dom'
import cx from 'classnames'
import { CSSTransition } from 'react-transition-group'
import { OverlayProps } from 'types/overlay'
import { noop, hasScrollBar, getScrollBarWidth } from '../utils/tool'
import Icon from '../Icon'

interface Animate {
  bodyRef: React.MutableRefObject<React.CSSProperties>
  wrapperRef: React.MutableRefObject<HTMLElement>
  maskRef: React.MutableRefObject<HTMLElement>
  destroy: boolean
}

const { useEffect } = React

const defaultProps: OverlayProps = {
  visible: false,
  hasMask: true,
  onClose: noop,
  maskClosable: true,
  destroy: true,
  prefixCls: 'snake-overlay',
  maskTimeout: 300,
  contentTimeout: 300,
  maskAnimation: 'fade',
  contentAnimation: 'zoom',
  closable: true,
  esc: true,
  afterClose: noop
}

const getBodyStyle = (bodyRef: React.MutableRefObject<React.CSSProperties>) => {
  const style: React.CSSProperties = {}
  if ('paddingRight' in document.body.style) {
    style.paddingRight = document.body.style.paddingRight
  }
  if ('overflow' in document.body.style) {
    style.overflow = document.body.style.overflow
  }
  bodyRef.current = style
}

const setDestroyStyle = ({ wrapperRef, maskRef }: Animate, display: string) => {
  wrapperRef.current.style.display = display
  maskRef.current.style.display = display
}

const onAnimateEnter = (animate: Animate) => {
  getBodyStyle(animate.bodyRef)
  if (hasScrollBar()) {
    // tslint:disable-next-line: radix
    document.body.style.paddingRight = `${parseInt(animate.bodyRef.current.paddingRight as string) +
      getScrollBarWidth()}px`
  }
  document.body.style.overflow = 'hidden'
  if (!animate.destroy) {
    setDestroyStyle(animate, '')
  }
}

const onAnimateExist = (animate: Animate) => {
  document.body.style.overflow = animate.bodyRef.current.overflow || ''
  document.body.style.paddingRight = `${animate.bodyRef.current.paddingRight}` || ''
  if (!animate.destroy) {
    setDestroyStyle(animate, 'none')
  }
}

const Overlay: React.FC<OverlayProps> = (overlay, ref) => {
  const props = { ...defaultProps, ...overlay }
  const {
    prefixCls,
    wrapperClassName,
    visible,
    destroy,
    hasMask,
    wrapperStyle,
    maskTimeout,
    contentTimeout,
    children,
    header,
    footer,
    maskAnimation,
    contentAnimation,
    zIndex,
    closable,
    maskClosable,
    onClose,
    esc,
    afterClose
  } = props

  const bodyRef = React.useRef<React.CSSProperties>()
  const wrapperRef = React.useRef<HTMLDivElement>()
  const maskRef = React.useRef()
  // 因为 动画库的原因，如果 `unmountExist` 为 false 就会直接渲染出来 不是很好
  const [firstTime, setFirstTime] = React.useState(true)

  if (visible) {
    if (firstTime) {
      setFirstTime(false)
    }
  }

  // 处理destroy为false的时候第一次没动画的问题
  const getUnmount = React.useCallback(() => {
    let unmountOnExit = true
    if (destroy) {
      unmountOnExit = true
    } else {
      if (firstTime) {
        unmountOnExit = true
      } else {
        unmountOnExit = false
      }
    }
    return unmountOnExit
  }, [firstTime, destroy])

  const getCls = React.useCallback(() => {
    const classStr = cx(prefixCls, wrapperClassName)
    return classStr
  }, [prefixCls, wrapperClassName])

  const getZIndex = React.useCallback(() => {
    const style: React.CSSProperties = {}
    if (zIndex) {
      style.zIndex = zIndex
    }
    return style
  }, [zIndex])

  const maskClick = (e: React.MouseEvent<HTMLElement>) => {
    if (e.target === e.currentTarget) {
      onClose(e)
    }
  }

  const keyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    if (esc && e.key === 'Escape') {
      e.stopPropagation()
      onClose(e as any)
    }
  }

  const renderMask = () => {
    const maskCls = `${prefixCls}-${maskAnimation}`
    const style = getZIndex()
    if (hasMask) {
      return (
        <CSSTransition
          in={visible}
          timeout={maskTimeout}
          classNames={maskCls}
          unmountOnExit={getUnmount()}
          appear
        >
          <div className={`${prefixCls}-mask`} style={style} ref={maskRef} />
        </CSSTransition>
      )
    }
    return null
  }

  const renderContent = () => {
    const contentCls = `${prefixCls}-${contentAnimation}`
    const style = getZIndex()
    return (
      <CSSTransition
        in={visible}
        timeout={contentTimeout}
        classNames={contentCls}
        unmountOnExit={getUnmount()}
        onEnter={() => onAnimateEnter({ bodyRef, wrapperRef, maskRef, destroy })}
        onExited={() => {
          afterClose()
          onAnimateExist({ bodyRef, wrapperRef, maskRef, destroy })
        }}
        appear
      >
        <div
          className={getCls()}
          onClick={maskClosable ? maskClick : undefined}
          onKeyDown={keyDown}
          ref={wrapperRef}
          style={style}
          tabIndex={-1}
        >
          <div className={`${prefixCls}-wrapper`} style={wrapperStyle} ref={ref}>
            {renderClosable()}
            {header ? <div className={cx(`${prefixCls}-header`)}>{header}</div> : null}
            <div className={cx(`${prefixCls}-wrapper-inner`)}>{children}</div>
            {footer ? <div className={cx(`${prefixCls}-footer`)}>{footer}</div> : null}
          </div>
        </div>
      </CSSTransition>
    )
  }

  const renderClosable = () => {
    if (closable) {
      return (
        <div className={`${prefixCls}-close`} onClick={onClose}>
          <Icon type="close" />
        </div>
      )
    }
    return null
  }

  const getContainer = () => {
    return (
      <>
        {renderMask()}
        {renderContent()}
      </>
    )
  }

  useEffect(() => {
    if (wrapperRef.current) {
      wrapperRef.current.focus()
    }
  }, [visible])

  return ReactDOM.createPortal(getContainer(), document.body)
}

export default React.forwardRef(Overlay)
