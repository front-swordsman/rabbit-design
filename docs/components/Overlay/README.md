---
imports:
  import OverlayDemoCode from '!raw-loader!./demo.tsx';
  import OverlayDemo from './demo.tsx';
---

# Overlay 弹出层

弹出框之类的选择层

## 简单示例

:::demo 
<Block code={OverlayDemoCode} des="弹出层"> 
<OverlayDemo /> 
</Block> 
:::

## Api

| 属性             | 说明                                                             | 类型               | 默认值          | 必填 |
| :--------------- | :--------------------------------------------------------------- | :----------------- | :-------------- | :--- |
| visible          | 是否可见                                                         | boolean            | false           | --   |
| hasMask          | 是否有遮罩层                                                     | boolean            | true            | --   |
| onClose          | 关闭弹层的回调                                                   | () => void         | noop            | --   |
| maskClosable     | 是否点击遮罩层关闭弹层                                           | boolean            | true            | --   |
| destroy          | 关闭弹层时，是否销毁该弹层                                       | boolean            | true            | --   |
| wrapperClassName | 包裹层的类名                                                     | string             | --              | --   |
| wrapperStyle     | 包裹层的样式                                                     | object             | --              | --   |
| maskAnimation    | 弹层的动画，可以参考`react-transition-group`                     | string             | `fade`          | --   |
| contentAnimation | 内容区的动画, 可以参考`react-transition-group`                   | string             | `zoom`          | --   |
| zIndex           | z轴方向的层级                                                    | number             | --              | --   |
| prefixCls        | 前缀                                                             | string             | `snake-overlay` | --   |
| maskTimeout      | 遮罩层动画的持续时间, 可以参考`react-transition-group`, 单位毫秒 | `number`, `object` | 300             | --   |
| contentTimeout   | 内容区的动画持续时间, 可以参考`react-transition-group`, 单位毫秒 | `number`, `object` | 300             | --   |
| header           | 是否有头部                                                       | ReactNode          | --              | --   |
| footer           | 是否有底部                                                       | ReactNode          | --              | --   |
| closable         | 是否显示右上角的叉号                                             | boolean            | `true`          | --   |