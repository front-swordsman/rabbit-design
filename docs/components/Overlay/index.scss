$modal: 'snake-overlay';
$modalIndex:999;
$border-radius-normal: 4px;

.#{$modal} {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  z-index: $modalIndex;
  display: flex;
  justify-content: center;
  align-items: center;
  outline: 0;

  &-hidden {
    display: none;
  }

  &-mask {
    position: fixed;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    background: rgba(0, 0, 0, 0.45);
    z-index: $modalIndex;
  }

  &-wrapper {
    background: #fff;
    border-radius: $border-radius-normal;
    width: 500px;
    position: relative;
  }

  &-header {
    padding: 16px 24px;
    border-bottom: 1px solid #e8e8e8;
  }

  &-close {
    position: absolute;
    right: 0;
    top: 0;
    color: rgba(0, 0, 0, 0.45);
    width: 55px;
    height: 55px;
    line-height: 55px;
    text-align: center;
    cursor: pointer;

    .uni-icon {
      font-size: 16px;
    }
  }

  &-body {
    &-contain {
      padding: 24px;
      font-size: 14px;
      line-height: 1.5;
      word-wrap: break-word;
    }
  }

  &-footer {
    padding: 10px 16px;
    text-align: right;
    border-top: 1px solid #e8e8e8;

    &-container {
      .uni-btn-button_btn {
        &:last-child {
          margin-left: 8px;
        }
      }
    }
  }

  &-fade-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: cubic-bezier(0.55, 0, 0.55, 0.2);
    animation-play-state: paused;
  }

  &-fade-enter-active {
    animation-name: modalFadeIn;
    animation-play-state: running;
  }

  &-fade-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: cubic-bezier(0.55, 0, 0.55, 0.2);
    animation-play-state: paused;
  }

  &-fade-exit-active {
    animation-name: modalFadeOut;
    animation-play-state: running;
  }

  &-zoom-enter {
    opacity: 0;
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: cubic-bezier(0.55, 0, 0.55, 0.2);
    animation-play-state: paused;
  }

  &-zoom-enter-active {
    animation-name: modalZoomIn;
    animation-play-state: running;
  }

  &-zoom-exit {
    animation-duration: 0.3s;
    animation-fill-mode: both;
    animation-timing-function: cubic-bezier(0.55, 0, 0.55, 0.2);
    animation-play-state: paused;
  }

  &-zoom-exit-active {
    animation-name: modalZoomOut;
    animation-play-state: running;
  }
}

@keyframes modalFadeIn {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes modalFadeOut {
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}

@keyframes modalZoomIn {
  0% {
    opacity: 0;
    transform: scale(0, 0);
  }

  100% {
    opacity: 1;
    transform: scale(1, 1);
  }
}

@keyframes modalZoomOut {
  0% {
    transform: scale(1, 1);
  }

  100% {
    opacity: 0;
    transform: scale(0, 0);
  }
}