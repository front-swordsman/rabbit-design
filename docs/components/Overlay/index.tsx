import * as React from 'react'
import OverlayDoc from './README.md'

export default function OverlayView() {
  return <OverlayDoc />
}
