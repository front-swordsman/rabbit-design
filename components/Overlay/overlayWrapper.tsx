import * as React from 'react'
import { createPortal } from 'react-dom'
import { CSSTransition } from 'react-transition-group'
import { OverlayProps } from 'types/overlay'
import Overlay from './overlay'

const noop = () => {}
const { useState } = React

const defaultProps = {
  prefixCls: 'snake-overlay',
  visible: false,
  mask: true,
  maskClosable: true,
  closable: false,
  autoFix: false,
  destroy: true,
  onClose: noop,
  maskAnimationName: 'fade',
  animationName: 'fade'
}

const setBodyStyle = (visible: boolean) => {
  document.body.style.overflow = visible ? 'hidden' : null
}

const OverlayWrapper: React.SFC<OverlayProps> = overlayWrap => {
  const props = { ...defaultProps, ...overlayWrap }
  const [firstTime, setFirstTime] = useState(true)
  const { visible, destroy, prefixCls } = props

  if (visible) {
    if (firstTime) {
      setFirstTime(false)
    }
  }

  // 处理destroy为false的时候第一次没动画的问题
  const getUnmount = () => {
    let props: any = {}
    if (destroy) {
      props.unmountOnExit = true
    } else {
      if (firstTime) {
        props.unmountOnExit = true
      } else {
        props.unmountOnExit = false
      }
    }
    return props
  }

  if (!destroy) {
    setBodyStyle(visible)
  }
  return createPortal(
    <CSSTransition in={visible} timeout={300} classNames={`${prefixCls}-fade`} {...getUnmount()}>
      <Overlay {...props} />
    </CSSTransition>,
    document.body
  )
}

export default OverlayWrapper
