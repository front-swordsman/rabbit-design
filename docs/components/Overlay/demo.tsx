import * as React from 'react'
import { Overlay, Button } from 'components'
import './index.scss'

export default function OverlayDemo() {
  const [visible, setVisible] = React.useState(false)
  const [visible1, setVisible1] = React.useState(false)
  return (
    <div>
      <h3>默认关闭弹窗销毁组件</h3>
      <Overlay visible={visible} onClose={() => setVisible(false)} zIndex={1020}>
        <div>内容区</div>
      </Overlay>
      <Button onClick={() => setVisible(true)}>点击显示弹出层</Button>
      <h3>关闭弹窗不销毁组件</h3>
      <Overlay visible={visible1} onClose={() => setVisible1(false)} destroy={false}>
        <div>内容区(不销毁组件)</div>
      </Overlay>
      <Button onClick={() => setVisible1(true)}>点击显示弹出层(不销毁组件)</Button>
    </div>
  )
}
