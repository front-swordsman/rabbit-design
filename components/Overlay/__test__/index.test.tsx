import * as React from 'react'
import 'jest-dom/extend-expect'
import { render, fireEvent } from 'react-testing-library'
import Overlay from '../index'

jest.mock('react-transition-group', () => {
  return {
    CSSTransition: props => {
      return <div>{props.in || !props.unmountOnExit ? props.children : null}</div>
    }
  }
})

describe('overlay test', () => {
  it('snapshot overlayHide', () => {
    const overlayHide = render(<Overlay visible={false}>内容区</Overlay>)
    expect(overlayHide.baseElement).toMatchSnapshot()
  })

  it('snapshot overlayShow', () => {
    const overlayShow = render(<Overlay visible>内容</Overlay>)
    expect(overlayShow.baseElement).toMatchSnapshot()
  })

  it('snapshot overlayHeader', () => {
    const overlayHeader = render(<Overlay header={<div>头部</div>} visible />)
    expect(overlayHeader.baseElement).toMatchSnapshot()
  })

  it('snapshot overlayFooter', () => {
    const overlayFooter = render(<Overlay footer={<div>底部</div>} visible />)
    expect(overlayFooter.baseElement).toMatchSnapshot()
  })

  it('snapshot overlayClosable', () => {
    const overlayClosable = render(
      <Overlay closable={false} visible>
        内容
      </Overlay>
    )
    expect(overlayClosable.baseElement).toMatchSnapshot()
  })

  it('portal in body', () => {
    const { getByText, rerender, queryByText } = render(
      <Overlay visible>
        <div>test</div>
      </Overlay>
    )
    expect(getByText('test')).toBeInTheDocument()
    rerender(
      <Overlay visible={false}>
        <div>test</div>
      </Overlay>
    )
    expect(queryByText('test')).not.toBeInTheDocument()
  })

  it('fire click', () => {
    function OverlayDemo() {
      const [visible, setVisible] = React.useState(false)
      return (
        <div>
          <Overlay visible={visible} onClose={() => setVisible(false)}>
            <div>test</div>
          </Overlay>
          <button onClick={() => setVisible(true)}>点击显示</button>
        </div>
      )
    }
    const { getByText, queryByText } = render(<OverlayDemo />)
    const btn = getByText(/点击显示/i)
    fireEvent.click(btn)
    expect(getByText('test')).toBeInTheDocument()
    const wrapperNode = queryByText('test').parentNode.parentNode.parentNode as HTMLElement
    fireEvent.click(wrapperNode)
    expect(queryByText('test')).not.toBeInTheDocument()
  })

  it('not destroy', () => {
    function OverlayDemo() {
      const [visible, setVisible] = React.useState(false)
      return (
        <div>
          <Overlay visible={visible} onClose={() => setVisible(false)} destroy={false}>
            <div>test</div>
          </Overlay>
          <button onClick={() => setVisible(true)}>点击显示</button>
        </div>
      )
    }
    const { getByText, queryByText } = render(<OverlayDemo />)
    const btn = getByText(/点击显示/i)
    fireEvent.click(btn)
    expect(getByText('test')).toBeInTheDocument()
    const wrapperNode = queryByText('test').parentNode.parentNode.parentNode as HTMLElement
    fireEvent.click(wrapperNode)
    expect(getByText(/test/i)).toBeInTheDocument()
  })

  it('no mask', () => {
    render(
      <Overlay hasMask={false} visible>
        内容区
      </Overlay>
    )
    const maskNode = document.querySelector('.snake-overlay-mask')
    expect(maskNode).not.toBeInTheDocument()
  })
})
