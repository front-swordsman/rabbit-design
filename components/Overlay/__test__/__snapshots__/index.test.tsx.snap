// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overlay test snapshot overlayClosable 1`] = `
<body>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
  <div />
  <div>
    <div
      class="snake-overlay-mask"
    />
  </div>
  <div>
    <div
      class="snake-overlay"
    >
      <div
        class="snake-overlay-wrapper"
      >
        <div
          class="snake-overlay-wrapper-inner"
        >
          内容
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`overlay test snapshot overlayFooter 1`] = `
<body>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
  <div />
  <div>
    <div
      class="snake-overlay-mask"
    />
  </div>
  <div>
    <div
      class="snake-overlay"
    >
      <div
        class="snake-overlay-wrapper"
      >
        <div
          class="snake-overlay-close"
        >
          <svg
            class="snake-icon snake-icon-close"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </div>
        <div
          class="snake-overlay-wrapper-inner"
        />
        <div
          class="snake-overlay-footer"
        >
          <div>
            底部
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`overlay test snapshot overlayHeader 1`] = `
<body>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
  <div />
  <div>
    <div
      class="snake-overlay-mask"
    />
  </div>
  <div>
    <div
      class="snake-overlay"
    >
      <div
        class="snake-overlay-wrapper"
      >
        <div
          class="snake-overlay-close"
        >
          <svg
            class="snake-icon snake-icon-close"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </div>
        <div
          class="snake-overlay-header"
        >
          <div>
            头部
          </div>
        </div>
        <div
          class="snake-overlay-wrapper-inner"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`overlay test snapshot overlayHide 1`] = `
<body>
  <div />
  <div />
  <div />
</body>
`;

exports[`overlay test snapshot overlayShow 1`] = `
<body>
  <div />
  <div>
    <div
      class="snake-overlay-mask"
    />
  </div>
  <div>
    <div
      class="snake-overlay"
    >
      <div
        class="snake-overlay-wrapper"
      >
        <div
          class="snake-overlay-close"
        >
          <svg
            class="snake-icon snake-icon-close"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </div>
        <div
          class="snake-overlay-wrapper-inner"
        >
          内容
        </div>
      </div>
    </div>
  </div>
  <script
    src="https://at.alicdn.com/t/font_1127944_82mztmm5t8t.js"
  />
</body>
`;
